Check for Edit permission of user while accessing Feature Flags

ajeety4 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR introduces functionality that restricts users from adding new domains to feature flags (FFs) if they do not have edit permissions.

The permission is managed via the ToggleEditPermission model, which grants access at the FF category (tag) level.The ability to assign this permission to users is not included in this PR and will follow in a subsequent update.

**Changes in this PR affect the following two pages:**

**1. Admin > Feature Flags**
- If a user does not have edit access to a feature flag, the Edit button is replaced with View.
- On the feature flag edit page, such users will not be able to add or remove domains/users — they will only be able to view them.
- An appropriate alert message is displayed to indicate this.

**User with Access**  (No change here)
![image](https://github.com/user-attachments/assets/bffcfba7-a464-410c-9407-cb7b7ad18a55)

![image](https://github.com/user-attachments/assets/bfdfa4c4-b680-41df-86e8-23e8a9b42fb3)


**User without Access**
![image](https://github.com/user-attachments/assets/23624680-e963-459c-9879-092b39a570ea)

![image](https://github.com/user-attachments/assets/40a4ac74-e67c-4b9a-86ac-2f7e77061ca8)


**2. Project Settings > Feature Flags and Privileges**
- Feature flags for which the user lacks access will have the Enabled/Not Enabled button disabled (greyed out).
- A help icon is displayed to indicate the reason for the restriction.

**User with Access**  (No change here)
![image](https://github.com/user-attachments/assets/686188ff-b68f-4931-ba85-cfeb2f852b00)

**User without Access**
Help icon when clicked opens the popup

![image](https://github.com/user-attachments/assets/c710234f-5c3d-4b26-971e-f3d87459c27e)


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[Ticket](https://dimagi.atlassian.net/browse/SC-4338)
[Spec](https://docs.google.com/document/d/1arXfsgW7psnlimYwXqN27mCujdhGCrk_xOm8IlZbKJk/edit?tab=t.0#heading=h.ytrlro455iap)

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
NA

## Safety Assurance

### Safety story
<!--
￼
Add a description
Comment
￼Write
￼Preview
￼￼￼￼￼￼
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested on local.
Affects internal superusers pages only. This PR will not change the current access level of the superusers.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
New tests included.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA pending.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

ajeety4 committed:
shortcut to get editable tags for a user

ajeety4 committed:
typo

ajeety4 committed:
caching

ajeety4 committed:
show view button if user does not have edit permission for toggle

ajeety4 committed:
show edit and save buttons only if user has access

ajeety4 committed:
check for edit permission while save

ajeety4 committed:
disable button if no edit access in project feature flag settings

ajeety4 committed:
add helps text for clarity on why button is disabled

ajeety4 committed:
check for edit toggle access in view from Project FF settings

ajeety4 committed:
edit flag page: disable input box if no edit access

ajeety4 committed:
edit flag page: show message for clarity if no edit access

ajeety4 committed:
allow all toggles on staging

ajeety4 committed:
test cases for checking access on admin feature flags

ajeety4 committed:
test cases for checking access on project > feature flags

ajeety4 committed:
resolves html syntax errors

ajeety4 committed:
prettify

ajeety4 committed:
sync bootstrap file

mkangia commented:
Hey @ajeety4 

Just tagging that this is on my list to review. I will do so this week.

mkangia commented on corehq/toggles/sql_models.py:
```diff
@@ -16,9 +17,11 @@ class ToggleEditPermission(models.Model):
         super().save(*args, **kwargs)
 
     @classmethod
+    @quickcache(['tag_slug'], timeout=60 * 60 * 24)
```
nit: Considering we are clearing it when needed, is there a need for this to expire at all? 24 hours seems fair, I am thinking if it could be cached until cleared.

ajeety4 replied:
Good point. Yes, I think we can store it for a longer duration.
There does not seem to be a direct way to set it to last forever (until cleared); however, we can store it for a long duration, effectively achieving the same result.
Glancing through the HQ codebase, it looks like we commonly use a `1 week` value for longer cache durations. 
How does `1 week` sound ?

ajeety4 replied:
Addressed in https://github.com/dimagi/commcare-hq/pull/36233/commits/7247fa6c82cc6e4cebae8d35bb2aa6a55a83bc32

mkangia replied:
1 week is better than 1 day :+1:

mkangia commented on corehq/toggles/shortcuts.py:
```diff
@@ -98,3 +98,16 @@ def find_domains_with_toggle_enabled(toggle):
     return [user[len(prefix):] for user in doc.enabled_users if user.startswith(prefix)]
+
+
+def get_tags_with_edit_permission(username):
```
nit: how about `get_editable_toggle_tags_for_user`?

ajeety4 replied:
https://github.com/dimagi/commcare-hq/pull/36233/commits/997e5f608f32c4b328e4709d5a40d200b709c042

mkangia commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -154,6 +154,10 @@ def static_toggle(self):
 
+    @property
+    def can_edit_toggle(self):
+        return self.static_toggle.tag in get_tags_with_edit_permission(self.request.user.username)
```
Can we add a way to check directly for this tag is user has edit access?
It seems unreasonable to find all editable tags.

ajeety4 replied:
https://github.com/dimagi/commcare-hq/pull/36233/commits/e495de38437dddf6c75825130c4d36572ea17ae1

mkangia commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -189,6 +189,9 @@ def page_context(self):
 
     def post(self, request, *args, **kwargs):
+        if not self.can_edit_toggle:
+            return HttpResponseForbidden("You do not have permission to edit this toggle.")
```
nit: is feature flag the term used in UI? and hence error messages?

mkangia commented on corehq/apps/domain/views/internal.py:
```diff
@@ -270,6 +270,8 @@ def _sort_key(toggle):
         return sorted(unsorted_toggles, key=_sort_key)
 
     def _editable_tags_slugs(self):
+        if settings.SERVER_ENVIRONMENT == 'staging':
```
Can you share the intent behind this? I don't think we need to do this for non prod environments. I am thinking about QA and also the queries getting skipped on staging which makes the environment different from prod which isn't good for investigations.

ajeety4 replied:
Quoting the intention from the spec
`it allows developers to test out all features on staging, which further allows them to figure out if a particular FF solves a problem or not without having to go through the churn of requesting permissions.`

I agree with the concern raised here, however the intent here does makes sense to me. 
One thing we can do is enable all toggles for all users on staging manually through shell. However any new user will have through go to support/dev to gain access to a toggle tag which seems like a bit of a barrier.

mkangia replied:
> However any new user will have through go to support/dev to gain access to a toggle tag which seems like a bit of a barrier

Its only tag related yes? if someone has access to a tag already, they won't be blocked if they add a new toggle to the tag right? So seems like a one time effort.

We can check with support how they do currently for superusers on staging? I think its quicker and does not need official approval.

I am hesitant to have a special behavior for an environment named "staging". If it was an environment setting one can set on staging & other test/dev environments, I'd feel better but for now, I don't see the need for it.

ajeety4 replied:
> ts only tag related yes? if someone has access to a tag already, they won't be blocked if they add a new toggle to the tag right?

Correct. We would need to provide access for new users only. 

> We can check with support how they do currently for superusers on staging?
Cool, will check on this.

ajeety4 replied:
Hey @mkangia , I checked with support, and the current process requires approval for superuser access to staging.
Since we are going to allow adding toggle (feature flag) edit access through the same UI as for adding superusers, I think this can be easily managed when providing superuser access to new users on staging.
For existing users, I will enable it through the shell for staging.

In light of this, the staging check was removed in fb3a66c82387644810db6f8dc55d034751e857ff

mkangia replied:
Cool, so the same list will be used for staging? seems reasonable.

ajeety4 replied:
> Cool, so the same list will be used for staging? 

For staging, all superusers will have access to all the toggles.

ajeety4 replied:
By `permissions`, if you mean `restrictions`-  Yes. We want to let have all superusers access on staging as per the [intent](https://github.com/dimagi/commcare-hq/pull/36233#discussion_r2062602176).

mkangia reviewed (changes requested):
looks good @ajeety4 

Just couple of quick points.

One thing about the message on the edit toggle page, that, it could get really hidden by being low on a page with enough domains/users added, so I suggest we move it up or wherever seems better.

ajeety4 commented on corehq/toggles/shortcuts.py:
```diff
@@ -98,3 +98,16 @@ def find_domains_with_toggle_enabled(toggle):
+    from corehq.toggles.sql_models import ToggleEditPermission
+
+    allowed_tags = []
+    for tag in ALL_TAGS:
```
This iterates through **tags** (and not toggles), which has around 10 items, so I think the performance here should be very fast. The database call inside the iteration is cached.Not iterating would be better; however, I think for practical purposes, the performance gain would be negligible given the small number of items in the iterator.

Sidenote: I did consider caching this method, but quickcache does not support clearing the cache for all parameters at once.So, I went ahead and cached the get object directly, which also seems safer given that the use case here is related to access/permission

ajeety4 committed:
rename

ajeety4 committed:
check user edit access for a tag instead of fetching all editable tags

ajeety4 committed:
nit:increase cache timeout

ajeety4 committed:
nit: rename error message

ajeety4 committed:
nit: move info message at top

mkangia commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -201,7 +201,7 @@ def page_context(self):
     def post(self, request, *args, **kwargs):
         if not self.can_edit_toggle:
-            return HttpResponseForbidden("You do not have permission to edit this toggle.")
+            return HttpResponseForbidden("You do not have permission to edit this Feature Flag.")
```
nit: Feature Flag or feature flag?

ajeety4 replied:
https://github.com/dimagi/commcare-hq/pull/36233/commits/65660c5c096f9598f0550b7f59270d670e2aaa77

mkangia reviewed (commented):
Changes look good :+1:

ajeety4 committed:
nit: casing

ajeety4 committed:
corrects condition

ajeety4 committed:
updates test case for rename and new method

ajeety4 committed:
toggle edit access for staging can be managed through UI

mkangia commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -406,7 +406,7 @@ def set_toggle(request, toggle_slug):
     if (
         settings.SERVER_ENVIRONMENT != 'staging'
-        and can_user_edit_tag(request.user.username, static_toggle.tag)
+        and not can_user_edit_tag(request.user.username, static_toggle.tag)
```
:-o

mkangia reviewed (commented):
Looks good.

Are we taking this through QA? I'd suggest we should.
We can split that into its own ticket to unblock dev ticket.

ajeety4 commented:
> Looks good.
> 
> Are we taking this through QA? I'd suggest we should. We can split that into its own ticket to unblock dev ticket.

Thanks Manish for the valuable comments on this PR.
As suggested, I have created a QA ticket for this.

mkangia commented:
Thanks @ajeety4 

I am happy with the changes but holding off on QA for approval.
Good luck.

ajeety4 committed:
permission entry is required for tag

ajeety4 committed:
updates test cases

ajeety4 committed:
Merge pull request #36334 from dimagi/ay/ff-permission-entry

Tag Permission entry is required for user to edit FFs

ajeety4 committed:
remove unused import

orangejenny committed:
Merge branch 'master' into ay/ff-access-ui

ajeety4 committed:
Merge branch 'master' into ay/ff-access-ui

ajeety4 committed:
updates the message if no permission

we no longer want to consider access requests as it would be managed entirely by SAAS

ajeety4 committed:
Merge branch 'master' into ay/ff-access-ui

ajeety4 committed:
add missed line during merge

mkangia reviewed (approved):
@ajeety4 
Confirming this is QA passed? the label says QA Awaiting.

ajeety4 committed:
migration to assign edit permissions to all superusers

ajeety4 committed:
Merge branch 'master' into ay/ff-access-ui

ajeety4 committed:
Merge pull request #36762 from dimagi/ay/ff-edit-access-migration

Simple migration to assigns edit permissions for all toggle tags to superusers to maintain consistency with the current behavior.

ajeety4 committed:
Merge branch 'master' into ay/ff-access-ui

ajeety4 committed:
Merge branch 'master' into ay/ff-access-ui

kaapstorm commented on corehq/toggles/migrations/0002_assign_permissions_superusers.py:
```diff
@@ -0,0 +1,41 @@
+def _assign_all_toggle_edit_permissions_to_superusers(apps, schema_editor):
+    superusers = User.objects.filter(is_superuser=True).values_list('username', flat=True)
+    for tag in ALL_TAGS:
+        toggle_permission = ToggleEditPermission.get_by_tag_slug(tag.slug)
```
I know these are cached with `quickcache()` but that still makes a network call. Much faster to cache these locally in this migration with a simple dictionary. e.g.

```python

# ...
    for tag in ALL_TAGS:
        toggle_permission = get_toggle_permission_by_tag_slug(tag.slug)
        # ...

toggle_permissions = {}

def get_toggle_permission_by_tag_slug(tag_slug):
    if tag_slug not in toggle_permissions:
        toggle_permission = (
            ToggleEditPermission.get_by_tag_slug(tag.slug)
            or ToggleEditPermission(tag_slug=tag.slug)
        )
        toggle_permissions[tag_slug] = toggle_permission
    return toggle_permissions[tag_slug]
```
